planning/mission_planner/src/mission_planner/mission_planner.cpp

```diff
@@ -334,19 +367,107 @@ void MissionPlanner::on_set_mrm_route(
   const SetMrmRoute::Service::Request::SharedPtr req,
   const SetMrmRoute::Service::Response::SharedPtr res)
 {
-  // TODO(Yutaka Shimizu): reroute for MRM
-  (void)req;
-  (void)res;
+  using ResponseCode = autoware_adapi_v1_msgs::srv::SetRoutePoints::Response;
+  change_state(RouteState::Message::CHANGING);
+
+  if (!planner_->ready()) {
+    change_state(RouteState::Message::SET);
+    throw component_interface_utils::ServiceException(
+      ResponseCode::ERROR_PLANNER_UNREADY, "The planner is not ready.");
+  }
+  if (!odometry_) {
+    change_state(RouteState::Message::SET);
+    throw component_interface_utils::ServiceException(
+      ResponseCode::ERROR_PLANNER_UNREADY, "The vehicle pose is not received.");
+  }
+  if (!normal_route_) {
+    change_state(RouteState::Message::SET);
+    throw component_interface_utils::ServiceException(
+      ResponseCode::ERROR_PLANNER_UNREADY, "Normal route is not set.");
+  }
+
+  // Plan route.
+  const auto new_route = create_route(req);
+
+  // check route safety
+  if (checkRerouteSafety(*normal_route_, new_route)) {
+    // success to reroute
+    change_mrm_route(new_route);
+    change_state(RouteState::Message::SET);
+    res->status.success = true;
+    return;
+  }
+
+  // Failed to reroute
+  if (mrm_route_) {
+    // if it has the old mrm route, use it
+    change_mrm_route(*mrm_route_);
+  } else {
+    // failed to go to mrm state, use normal route
+    change_route(*normal_route_);
+  }
+  change_state(RouteState::Message::SET);
+  res->status.success = false;
 }
 
 // NOTE: The route interface should be mutually exclusive by callback group.
 void MissionPlanner::on_clear_mrm_route(
-  const ClearMrmRoute::Service::Request::SharedPtr req,
+  [[maybe_unused]] const ClearMrmRoute::Service::Request::SharedPtr req,
   const ClearMrmRoute::Service::Response::SharedPtr res)
 {
-  // TODO(Yutaka Shimizu): reroute for MRM
-  (void)req;
-  (void)res;
+  using ResponseCode = autoware_adapi_v1_msgs::srv::SetRoutePoints::Response;
+  change_state(RouteState::Message::CHANGING);
```

Comment:
change_state is done after all guard conditions.

Reply:
done https://github.com/autowarefoundation/autoware.universe/pull/4027/commits/41357dc9a761fabd82eeb1346a74baad0bbc89c9